python/cudf_polars/cudf_polars/experimental/statistics.py

```diff
@@ -0,0 +1,229 @@
+        for name in left.schema
+        if name not in kstats
+    }
+    suffix = ir.options[3]
+    jstats |= {
+        name if name not in jstats else f"{name}{suffix}": right_column_stats.get(
+            name, ColumnStats(name=name)
+        )
+        for name in right.schema
+        if name not in kstats
+    }
+    return kstats | jstats
+
+
+@extract_base_stats.register(GroupBy)
+def _(
+    ir: GroupBy, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    (child,) = ir.children
+    child_column_stats = stats.column_stats.get(child, {})
+
+    # Update set of source columns we may lazily sample
+    _update_unique_stats_columns(
+        child_column_stats, [n.name for n in ir.keys], config_options
+    )
+
+    return {
+        n.name: child_column_stats.get(n.name, ColumnStats(name=n.name))
+        for n in ir.keys
+    } | {n.name: ColumnStats(name=n.name) for n in ir.agg_requests}
+
+
+@extract_base_stats.register(HStack)
+def _(
+    ir: HStack, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    (child,) = ir.children
+    child_column_stats = stats.column_stats.get(child, {})
+    new_cols = {
+        n.name: child_column_stats.get(n.value.name, ColumnStats(name=n.name))
+        if isinstance(n.value, expr.Col)
+        else ColumnStats(name=n.name)
+        for n in ir.columns
+    }
+    return child_column_stats | new_cols
+
+
+@extract_base_stats.register(Select)
+def _(
+    ir: Select, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    (child,) = ir.children
+    child_column_stats = stats.column_stats.get(child, {})
+    return {
+        n.name: child_column_stats.get(n.value.name, ColumnStats(name=n.name))
+        if isinstance(n.value, expr.Col)
+        else ColumnStats(name=n.name)
+        for n in ir.exprs
+    }
+
+
+@extract_base_stats.register(HConcat)
+def _(
+    ir: HConcat, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    return dict(
+        itertools.chain.from_iterable(
+            stats.column_stats.get(c, {}).items() for c in ir.children
+        )
+    )
+
+
+@extract_base_stats.register(Union)
+def _(
+    ir: Union, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    # TODO: We can preserve matching source statistics
+    return {name: ColumnStats(name=name) for name in ir.schema}
+
+
+def _extract_base_stats_preserve(
+    ir: IR, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    (child,) = ir.children
+    child_column_stats = stats.column_stats.get(child, {})
+    return {
+        name: child_column_stats.get(name, ColumnStats(name=name)) for name in ir.schema
+    }
+
+
+extract_base_stats.register(Cache, _extract_base_stats_preserve)
+extract_base_stats.register(Projection, _extract_base_stats_preserve)
+extract_base_stats.register(Sort, _extract_base_stats_preserve)
```

Comment:
You can register this for multiple types.

```suggestion
@extract_base_stats.register(Cache)
@extract_base_stats.register(Projection)
@extract_base_stats.register(Sort)
def _extract_base_stats_preserve(
    ir: Cache | Projection | Sort, stats: StatsCollector, config_options: ConfigOptions
) -> dict[str, ColumnStats]:
    (child,) = ir.children
    child_column_stats = stats.column_stats.get(child, {})
    return {
        name: child_column_stats.get(name, ColumnStats(name=name)) for name in ir.schema
    }
```